src/util/resourceValidationUtils.js

```diff
@@ -0,0 +1,124 @@
+        const validatedProfiles = new Set(req.body.meta.profile);
+        profiles.forEach(profile => {
+          validatedProfiles.add(profile);
+        });
+        req.body.meta.profile = Array.from(validatedProfiles);
+      } else {
+        req.body['meta'] = { profile: profiles };
+      }
+
+      next();
+    } else {
+      res.status(validationInfo.code).json(validationInfo.data);
+    }
+  } else {
+    next();
+  }
+}
+
+/**
+ * Determines the profiles the submitted resource is expected to be validated against
+ * @param {String} originalUrl the url the request was sent to
+ * @param {Object} body an express request body containing a FHIR resource
+ * @returns String or Array
+ */
+function retrieveProfiles(originalUrl, body) {
+  const profiles = [];
+  const params = originalUrl.split('/');
+
+  const metaProfiles = body?.meta?.profile;
+  if (metaProfiles) {
+    profiles.push(...metaProfiles);
+  }
+  if (params[params.length - 1] === '$submit-data') {
+    profiles.push('Parameters');
+  }
+  // We don't need to validate posted Parameters bodies for dollar-sign operations since these aren't stored in the db
+  else if (params[params.length - 1][0] === '$') {
+    return 'Calculation';
+  }
+  // Only param was base_version, so this is a transaction bundle upload
+  else if (params.length === 2) {
+    profiles.push('Bundle');
+  } else if (params.length > 2) {
+    /*
+     * If two or more params and not dollar-sign operation, third param must be the resourceType
+     * keep in mind, the first param will always be the empty string since originalUrl starts with '/'
+     */
+    profiles.push(params[2]);
+  }
+  return profiles;
+}
+
+/**
+ * Queries the validation server and returns an object with the validity outcome and, upon failure,
+ * information about what was invalid
+ * @param {String} validationUrl the url to query for resource validation
+ * @param {Object} body the body of an express request containing a FHIR resource
+ * @returns Object
+ */
+async function getValidationInfo(profiles, body, base_version) {
+  const validationUrl = `http://${process.env.VALIDATOR_HOST}:${
+    process.env.VALIDATOR_PORT
+  }/validate?profile=${profiles.join(',')}`;
+  let response;
+  try {
+    response = await axios.post(validationUrl, body);
+  } catch (e) {
+    /*
+        We can't throw server errors inside this middleware, so best to construct
+        our own operation outcomes and return them
+    */
+    const outcome = {};
+    outcome.id = uuidv4();
+    outcome.issue = [
+      {
+        severity: 'error',
+        code: 'exception',
+        details: {
+          text: `Validation server failed while validating against profiles: ${profiles.join(',')}`
+        }
+      }
+    ];
+    const OperationOutcome = resolveSchema(base_version, 'operationoutcome');
+    return { isValid: false, code: 500, data: new OperationOutcome(outcome).toJSON() };
+  }
+  if (response.data.issue[0].details.text !== 'All OK') {
```

Comment:
I think we should be a bit more generous here with what classifies validation being fine. There are cases where the validator will issue "informational" or "warning" level "issues", which I don't think will classify as "Bad Request". I think in later tasking we can think about how if at all we want to send those warnings back, but for now I think we should let them through.

A way to replicate this behavior is to change the URL of a code system on some coding (e.g. I changed a Patient resource's `identifier[0].type.coding[0].system` to be `http://example.com`. You'll see that we don't get "All Ok", but we also don't get anything with severity "error".

All this is to say I think we should make a helper that takes in the validation response and returns a bool for if it should pass or not. I think right now it should return true if the issue array contains _no_ objects with severity `error`, and false if any are present.

Let's unit test this baby in isolation too. It might change later